src/core/heavyGraph.js

```diff
@@ -0,0 +1,122 @@
+// @flow
+import {
+  type Edge,
+  type Node,
+  type EdgeAddressT,
+  type NodeAddressT,
+  Graph,
+} from "./graph";
+import {
+  type NodeWeight,
+  type EdgeWeight,
+  type WeightsComparison,
+} from "./weights";
+
+/**
+ * An entry in the array returned by AddressModule.toParts()
+ */
+export type AddressPart = string;
+
+/**
+ * the core structure in the Node weight-resolution tree
+ */
+export type NodeWeightBranch = {
+  +weight?: NodeWeight,
+  +children: Map<AddressPart, NodeWeightBranch>,
+};
+
+/**
+ * the core structure in the Edge weight-resolution tree
+ */
+export type EdgeWeightBranch = {
+  +weight?: EdgeWeight,
+  +children: Map<AddressPart, EdgeWeightBranch>,
+};
+
+/**
+ * The HeavyGraph class is a replacement for the WeightedGraph type. This class
+ * eliminates the need give every node and edge that exists a weight. Rather,
+ * address prefixes can be assigned a weight and when a node or edge weight is
+ * queried, the longest prefix with a weight assigned will have its value
+ * returned.
+ *
+ * Over the longer-term, WeightedGraph should be
+ * refactored out of existence HeavyGraph should completely replace it.
+ */
+export class HeavyGraph {
+  _nodeWeights: NodeWeightBranch;
+  _edgeWeights: EdgeWeightBranch;
+  graph: Graph;
+
+  /**
+   * Set the weight for a given "class" of nodes. Each node that contains the
+   * supplied prefix in its address will potentially return this weight if no
+   * weight is assigned for a larger prefix that the node's address contains.
+   */
+  setNodePrefixWeight(
+    _unused_prefix: NodeAddressT,
+    _unused_w: NodeWeight
+  ): this {
+    throw new Error("method not implemented");
+  }
+
+  /**
+   * Set the weights for a given "class" of edges. Each edge that contains the
+   * supplied prefix in its address will potentially return these weights if
+   * none are assigned for a larger prefix that the edge's address contains.
+   */
+  setEdgePrefixWeight(
+    _unused_prefix: EdgeAddressT,
+    _unused_w: NodeWeight
+  ): this {
+    throw new Error("method not implemented");
+  }
+
+  /**
+   * Add a node and optionally assign its weight. If a prefix of the node's
+   * address doesn't have a weight assigned, a weight must be included or an
+   * error will throw.
+   */
+  addNode(_unused_n: Node, _unused_w?: NodeWeight): this {
+    throw new Error("method not implemented");
+  }
+
+  /**
+   * Add an edge and optionally assign its weight. If a prefix of the edge's
+   * address doesn't have a weight assigned, a weight must be included or an
+   * error will throw.
+   */
+  addEdge(_unused_e: Edge, _unused_w?: EdgeWeight): this {
+    throw new Error("method not implemented");
+  }
+
+  /**
+   * Return the weight of the node, or the weight of the longest prefix of the
+   * node address if the node doesn't have its own weight.
+   */
+  getNodeWeight(_unused_a: NodeAddressT): NodeWeight {
+    throw new Error("method not implemented");
+  }
```

Comment:
Assuming we keep the default weights of `1` and `{1,1}` (with multiplicative semantics), as discussed in a comment above, then we can realize that weights are really associated with addresses and not with nodes. Since every node has an address, whenever we are retrieving a particular node we can return it with the weight associated with its address. 

But every address has a weight, and not every address has a node. So for the docstring to say "Return the weight of the node, or the weight of the address if the node doesn't have a weight" makes a confusing API. I think we should just create a clear and locally consistent API, where we are returning the weight associated with the address directly.